interfaces/connection.go

```diff
@@ -0,0 +1,250 @@
+	return nil
+}
+
+// Interface returns the name of the interface for this slot.
+func (slot *ConnectedSlot) Interface() string {
+	return slot.slotInfo.Interface
+}
+
+// Name returns the name of this slot.
+func (slot *ConnectedSlot) Name() string {
+	return slot.slotInfo.Name
+}
+
+// Snap returns the snap Info of this slot.
+func (slot *ConnectedSlot) Snap() *snap.Info {
+	return slot.slotInfo.Snap
+}
+
+// Apps returns all the apps associated with this slot.
+func (slot *ConnectedSlot) Apps() map[string]*snap.AppInfo {
+	return slot.slotInfo.Apps
+}
+
+// Hooks returns all the hooks associated with this slot.
+func (slot *ConnectedSlot) Hooks() map[string]*snap.HookInfo {
+	return slot.slotInfo.Hooks
+}
+
+// SecurityTags returns the security tags for this slot.
+func (slot *ConnectedSlot) SecurityTags() []string {
+	return slot.slotInfo.SecurityTags()
+}
+
+// StaticAttr returns a static attribute with the given key, or error if attribute doesn't exist.
+func (slot *ConnectedSlot) StaticAttr(key string) (interface{}, error) {
+	if val, ok := slot.staticAttrs[key]; ok {
+		return val, nil
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+// StaticAttrs returns all static attributes.
+func (slot *ConnectedSlot) StaticAttrs() map[string]interface{} {
+	return slot.staticAttrs
+}
+
+// Attr returns a dynamic attribute with the given name. It falls back to returning static
+// attribute if dynamic one doesn't exist. Error is returned if neither dynamic nor static
+// attribute exist.
+func (slot *ConnectedSlot) Attr(key string) (interface{}, error) {
+	if slot.dynamicAttrs != nil {
+		if val, ok := slot.dynamicAttrs[key]; ok {
+			return val, nil
+		}
+	}
+	return slot.StaticAttr(key)
+}
+
+// Attrs returns all dynamic attributes of this slot.
+func (slot *ConnectedSlot) Attrs() map[string]interface{} {
+	if slot.dynamicAttrs == nil {
+		slot.dynamicAttrs = make(map[string]interface{})
+	}
+	return slot.dynamicAttrs
+}
+
+// SetAttr sets the given dynamic attribute. Error is returned if the key is already used by a static attribute.
+func (slot *ConnectedSlot) SetAttr(key string, value interface{}) error {
+	if _, ok := slot.staticAttrs[key]; ok {
+		return fmt.Errorf("cannot change attribute %q as it was statically specified in the snap details", key)
+	}
+	if slot.dynamicAttrs == nil {
+		slot.dynamicAttrs = make(map[string]interface{})
+	}
+	// make a normalized copy
+	slot.dynamicAttrs[key] = copyRecursive(value)
+	return nil
+}
+
+// Interface returns the name of the interface for this connection.
+func (conn *Connection) Interface() string {
+	return conn.plug.plugInfo.Interface
+}
+
+func copyAttributes(value map[string]interface{}) map[string]interface{} {
+	return copyRecursive(value).(map[string]interface{})
+}
+
+func copyRecursive(value interface{}) interface{} {
+	// note: ensure all the mutable types (or types that need a conversion)
+	// are handled here.
+	switch v := value.(type) {
+	case int:
```

Comment:
The int and float cases probably deserve a comment about why this is done. The function seems to conflate JSON support and recursive copying.